File: src/tools/cash-flow-tools.ts
import { Tool, ToolResult } from '../types/index.js';
import { PythonBridge } from '../utils/python-bridge.js';

const pythonBridge = new PythonBridge();

const cashFlowTools: Tool[] = [
  {
    name: "cash_flow_statement",
    description: "Generate cash flow statement for specified period",
    inputSchema: {
      type: "object",
      properties: {
        startDate: { type: "string", format: "date" },
        endDate: { type: "string", format: "date" }
      },
      required: ["startDate", "endDate"]
    },
    handler: async (args: any): Promise<ToolResult> => {
      try {
        const result = await pythonBridge.callPythonFunction({
          module: 'cash_flow_tools',
          function: 'generate_cash_flow_statement',
          args: [args.startDate, args.endDate]
        });
        return result;
      } catch (error) {
        return {
          success: false,
          error: error instanceof Error ? error.message : String(error)
        };
      }
    }
  },

  {
    name: "cash_flow_forecast",
    description: "Forecast future cash flows with scenario analysis",
    inputSchema: {
      type: "object",
      properties: {
        monthsAhead: { type: "number", default: 12 },
        scenarios: {
          type: "object",
          properties: {
            base: { type: "number", default: 1.0 },
            conservative: { type: "number", default: 0.9 },
            optimistic: { type: "number", default: 1.1 }
          }
        }
      }
    },
    handler: async (args: any): Promise<ToolResult> => {
      try {
        const result = await pythonBridge.callPythonFunction({
          module: 'cash_flow_tools',
          function: 'forecast_cash_flow',
          args: [args.monthsAhead || 12, args.scenarios]
        });
        return result;
      } catch (error) {
        return {
          success: false,
          error: error instanceof Error ? error.message : String(error)
        };
      }
    }
  },

  {
    name: "cash_burn_analysis",
    description: "Analyze cash burn rate and calculate runway",
    inputSchema: {
      type: "object",
      properties: {
        monthsBack: { type: "number", default: 6 }
      }
    },
    handler: async (args: any): Promise<ToolResult> => {
      try {
        const result = await pythonBridge.callPythonFunction({
          module: 'cash_flow_tools',
          function: 'cash_burn_analysis',
          args: [args.monthsBack || 6]
        });
        return result;
      } catch (error) {
        return {
          success: false,
          error: error instanceof Error ? error.message : String(error)
        };
      }
    }
  },

  {
    name: "working_capital_analysis",
    description: "Analyze working capital changes and efficiency",
    inputSchema: {
      type: "object",
      properties: {
        startDate: { type: "string", format: "date" },
        endDate: { type: "string", format: "date" }
      },
      required: ["startDate", "endDate"]
    },
    handler: async (args: any): Promise<ToolResult> => {
      try {
        const result = await pythonBridge.callPythonFunction({
          module: 'cash_flow_tools',
          function: 'working_capital_analysis',
          args: [args.startDate, args.endDate]
        });
        return result;
      } catch (error) {
        return {
          success: false,
          error: error instanceof Error ? error.message : String(error)
        };
      }
    }
  },

  {
    name: "cash_flow_at_risk",
    description: "Calculate Cash Flow at Risk (CFaR) metric",
    inputSchema: {
      type: "object",
      properties: {
        confidenceLevel: { type: "number", default: 0.95, minimum: 0.01, maximum: 0.99 }
      }
    },
    handler: async (args: any): Promise<ToolResult> => {
      try {
        const result = await pythonBridge.callPythonFunction({
          module: 'cash_flow_tools',
          function: 'cash_flow_at_risk',
          args: [args.confidenceLevel || 0.95]
        });
        return result;
      } catch (error) {
        return {
          success: false,
          error: error instanceof Error ? error.message : String(error)
        };
      }
    }
  },

  {
    name: "liquidity_analysis",
    description: "Analyze current liquidity position and requirements",
    inputSchema: {
      type: "object",
      properties: {}
    },
    handler: async (): Promise<ToolResult> => {
      try {
        const result = await pythonBridge.callPythonFunction({
          module: 'cash_flow_tools',
          function: 'liquidity_analysis',
          args: []
        });
        return result;
      } catch (error) {
        return {
          success: false,
          error: error instanceof Error ? error.message : String(error)
        };
      }
    }
  }
];

export { cashFlowTools };

File: src/python/cash_flow_tools.py
import pandas as pd
import numpy as np
from datetime import datetime, date, timedelta
from typing import Dict, List, Optional, Any
import json

def generate_cash_flow_statement(start_date: str, end_date: str) -> Dict[str, Any]:
    """Generate cash flow statement for period"""
    # This would integrate with the CashFlowAnalyzer
    from cash_flow_analysis import CashFlowAnalyzer
    
    analyzer = CashFlowAnalyzer()
    # In practice, would load data from storage
    
    start_dt = datetime.strptime(start_date, '%Y-%m-%d').date()
    end_dt = datetime.strptime(end_date, '%Y-%m-%d').date()
    
    return analyzer.generate_cash_flow_statement(start_dt, end_dt)

def forecast_cash_flow(months_ahead: int = 12, scenarios: Optional[Dict[str, float]] = None) -> List[Dict]:
    """Forecast cash flows with scenario analysis"""
    from cash_flow_analysis import CashFlowAnalyzer
    
    analyzer = CashFlowAnalyzer()
    result = analyzer.forecast_cash_flow(months_ahead, scenarios)
    
    return result.to_dict('records') if hasattr(result, 'to_dict') else []

def cash_burn_analysis(months_back: int = 6) -> Dict[str, Any]:
    """Analyze cash burn rate and runway"""
    from cash_flow_analysis import CashFlowAnalyzer
    
    analyzer = CashFlowAnalyzer()
    return analyzer.cash_burn_analysis(months_back)

def working_capital_analysis(start_date: str, end_date: str) -> Dict[str, Any]:
    """Analyze working capital changes"""
    from cash_flow_analysis import CashFlowAnalyzer
    
    analyzer = CashFlowAnalyzer()
    start_dt = datetime.strptime(start_date, '%Y-%m-%d').date()
    end_dt = datetime.strptime(end_date, '%Y-%m-%d').date()
    
    return analyzer.working_capital_analysis(start_dt, end_dt)

def cash_flow_at_risk(confidence_level: float = 0.95) -> Dict[str, Any]:
    """Calculate Cash Flow at Risk (CFaR)"""
    from cash_flow_analysis import CashFlowAnalyzer
    
    analyzer = CashFlowAnalyzer()
    return analyzer.cash_flow_at_risk(confidence_level)

def liquidity_analysis() -> Dict[str, Any]:
    """Analyze liquidity position"""
    from cash_flow_analysis import CashFlowAnalyzer
    
    analyzer = CashFlowAnalyzer()
    return analyzer.liquidity_analysis()

if __name__ == "__main__":
    # Test cash flow tools
    print("Cash Flow Statement:")
    stmt = generate_cash_flow_statement('2024-01-01', '2024-12-31')
    print(json.dumps(stmt, indent=2, default=str))